Do not clone audio RTP packets

jmillan opened the pull request:
We are currently cloning all RTP packets and creating a shared_ptr out
of them.

* Do not clone audio RTP packets, saving the corresponding RtpPacket
  instance creation, memory allocation, memory copy, etc.

jmillan committed:
Do not clone audio RTP packets

We are currently cloning all RTP packets and creating a shared_ptr out
of them.

* Do not clone audio RTP packets, saving the corresponding RtpPacket
  instance creation, memory allocation, memory copy, etc.

ibc commented on worker/src/RTC/Router.cpp:
```diff
@@ -780,24 +780,27 @@ namespace RTC
+		// The packet does not need to be accessible in the future, do not clone it.
+		if (producer->GetKind() == RTC::Media::Kind::AUDIO)
+			std::shared_ptr<RTC::RtpPacket> sharedPacket(packet);
+		else
```
Brackets in if and else please

ibc reviewed (commented):
Looking good

jmillan committed:
handle feedback: lint

nazar-pc commented on worker/src/RTC/Router.cpp:
```diff
@@ -780,24 +780,31 @@ namespace RTC
+		// The packet does not need to be accessible in the future, do not clone it.
+		if (producer->GetKind() == RTC::Media::Kind::AUDIO)
+		{
+			std::shared_ptr<RTC::RtpPacket> sharedPacket(packet);
```
I don't think it does what you think it does, this is probably more correct:
```suggestion
			sharedPacket.reset(packet);
```

But it shouldn't compile because it is just a pointer and can't be de-allocated.

jmillan replied:
Yep, it's been fixed in the last commit.

nazar-pc commented on worker/src/RTC/Router.cpp:
```diff
@@ -780,24 +780,31 @@ namespace RTC
 		{
 			// Clone the packet so it holds its own buffer, usable for future
 			// retransmissions.
 			std::shared_ptr<RTC::RtpPacket> sharedPacket(packet->Clone());
```
```suggestion
			sharedPacket.reset(packet->Clone());
```

nazar-pc commented on worker/src/RTC/Transport.cpp:
```diff
@@ -1726,7 +1727,9 @@ namespace RTC
 		}
 
-		delete packet;
+		// Media forwarded, the packet lifetime is already handled.
+		if (result != RTC::Producer::ReceiveRtpPacketResult::MEDIA_FORWARDED)
+			delete packet;
```
This is super fragile, let's try to avoid such things when de-allocation of something depends on return value somewhere in completely different place.

jmillan replied:
Handled in the last commit. Sorry @nazar-pc, indeed even though I marked it as PR I already told @ibc to wait for review until a missing commit was pushed..., I forgot to label this as draft again..

nazar-pc reviewed (commented):
I don't think this actually works for audio at all, if I read it correctly audio packets with this change are always `nullptr`, which in Rust wouldn't compile or would crash immediately. This needs a different approach, probably a different `SendRtpPacket`, one for cloned packet in smart pointer and another with just raw pointer or something like that.

Though generally one clone per packet wasn't a bottleneck from my testing, but I agree it might be handy to avoid it if possible.

jmillan committed:
Transport: Do not delete the incoming RtpPacket

It will be deleted by Producer in case the packet is not forwared and
by Router otherwise.

jmillan commented:
> This needs a different approach, probably a different SendRtpPacket, one for cloned packet in smart pointer and another with just raw pointer or something like that.

Yes, this will be a next action.

jmillan committed:
cosmetic

jmillan commented:
Note for reviewers. This is a very small PR. Please review it in bulk rather than commit by commit since some things were re-done.

jmillan commented:
Converted to draft. Please, do not review yet.

jmillan commented:
> SendRtpPacket, one for cloned packet in smart pointer and another with just raw pointer or something like that.

This is what we had in mind for a second iteration. I'm giving it a try indeed.

ibc commented on worker/src/RTC/Producer.cpp:
```diff
@@ -621,15 +624,13 @@ namespace RTC
 
-			return ReceiveRtpPacketResult::DISCARDED;
+			result = ReceiveRtpPacketResult::DISCARDED;
+			goto done;
```
OMG really? Pretty sure we can avoid this...

I strongly prefer early return result.

jmillan replied:
not used in new PR

ibc commented:
THIS PR HAS BEEN CLOSED
???

jmillan commented:
> THIS PR HAS BEEN CLOSED

This, this PR has been closed in favour of https://github.com/versatica/mediasoup/pull/850
